### PR TITLE
Add NVIDIA Cosmos

### LINKED ISSUE #35565: Add cosmos from Nvidia
### Model description

https://www.nvidia.com/en-us/ai/cosmos/ on model is autoregressive, all open source

We might have a PR coming directly from NVIDIA! 

### Open source status

- [X] The model implementation is available
- [X] The model weights are available

### Provide useful links for the implementation

_No response_

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/35565

Conversion and generation with video only model is tested on 4B checkpoint. Next steps:
- [x] adapt conversion for 11B and 12B models if needed
- [x] verify that video-text generation works as well
- [x] match logits
- [x] add tests

### CODE REVIEW

**Comment by zucchini-nlp** [comment]
Ready for review. Note that most parts of the model are similar to existing Emu3 but we cannot copy with modular. Modular can't handle cases with `if/else` or for loops. Thus it was rewritten, making the modular file huge 🥲

File: src/transformers/models/cosmos/modular_cosmos.py [hunk 1]
```diff
@@ -0,0 +1,2297 @@
...
+        embed_dim: int = 6,
+        latent_channels: int = 16,
+        temporal_downsample_factor: int = 8,
+        attn_resolutions: List[int] = None,
+        base_channels: int = 128,
+        channel_multiplier: List[int] = [2, 4, 4],
+        num_res_blocks: int = 2,
+        hidden_size: int = 512,
+        patch_size: int = 4,
+        levels: List[int] = [8, 8, 8, 5, 5, 5],
```
Comment (ArthurZucker) [inline src/transformers/models/cosmos/modular_cosmos.py:141]: list defaults are never good!

File: src/transformers/models/cosmos/modular_cosmos.py [hunk 2]
```diff
@@ -0,0 +1,2297 @@
...
+        encoder_attentions (`tuple(torch.FloatTensor)`, *optional*, returned when `output_attentions=True` is passed or when `config.output_attentions=True`):
+            Tuple of `torch.FloatTensor` (one for each layer) of shape `(batch_size, num_heads, sequence_length,
+            sequence_length)`.
+
+            Attentions weights of the encoder, after the attention softmax, used to compute the weighted average in the
+            self-attention heads.
+    """
+
+
+class CosmosVQVAEVectorQuantizer(nn.Module):
```
Comment (ArthurZucker) [inline src/transformers/models/cosmos/modular_cosmos.py:472]: as we have other VQVAEs now in transformers make sur to standardize here, checking if we can't re-use / re-factor old blocks! 🤗

File: src/transformers/models/cosmos/modular_cosmos.py [hunk 3]
```diff
@@ -0,0 +1,2297 @@
...
+
+
+class CosmosVQVAEDecoderUpsample(nn.Module):
+    def __init__(self, in_channels, temporal_up: bool = True):
+        super().__init__()
+        self.conv1 = (
+            CosmosCausalConv3d(in_channels, in_channels, kernel_size=(3, 1, 1), stride=1, time_stride=1, padding=0)
+            if temporal_up
+            else nn.Identity()
+        )
```
Comment (ArthurZucker) [inline src/transformers/models/cosmos/modular_cosmos.py:622]: we should use if else!

File: src/transformers/models/cosmos/modular_cosmos.py [hunk 4]
```diff
@@ -0,0 +1,2297 @@
...
+        hidden_hlh = F.conv3d(
+            hidden_high_low, wavelets_high.unsqueeze(2).unsqueeze(3), groups=seq_len, stride=(1, 1, 2)
+        )
+        hidden_hhl = F.conv3d(
+            hidden_high_high, wavelets_low.unsqueeze(2).unsqueeze(3), groups=seq_len, stride=(1, 1, 2)
+        )
+        hidden_hhh = F.conv3d(
+            hidden_high_high, wavelets_high.unsqueeze(2).unsqueeze(3), groups=seq_len, stride=(1, 1, 2)
+        )
```
Comment (ArthurZucker) [inline src/transformers/models/cosmos/modular_cosmos.py:725]: wow that's ugly 🤣 we can probably make this simpler !

File: src/transformers/models/cosmos/modular_cosmos.py [hunk 5]
```diff
@@ -0,0 +1,2297 @@
...
+        dtype = hidden_states.dtype
+        wavelets = self.wavelets
+
+        groups = hidden_states.shape[1] // 8  # split into 8 spatio-temporal filtered tenors.
+        hl = wavelets.flip([0]).reshape(1, 1, -1).repeat([groups, 1, 1]).to(dtype=dtype)
+        hh = (wavelets * ((-1) ** self._arange)).reshape(1, 1, -1).repeat(groups, 1, 1).to(dtype=dtype)
+        xlll, xllh, xlhl, xlhh, xhll, xhlh, xhhl, xhhh = torch.chunk(hidden_states, 8, dim=1)
+
+        # Height height transposed convolutions.
+        xll = F.conv_transpose3d(xlll, hl.unsqueeze(2).unsqueeze(3), groups=groups, stride=(1, 1, 2))
```
Comment (ArthurZucker) [inline src/transformers/models/cosmos/modular_cosmos.py:764]: [..., None, None] is already more lightweight!

File: src/transformers/models/cosmos/modular_cosmos.py [hunk 6]
```diff
@@ -0,0 +1,2297 @@
...
+        xll += F.conv_transpose3d(xllh, hh.unsqueeze(2).unsqueeze(3), groups=groups, stride=(1, 1, 2))
```
Comment (ArthurZucker) [inline src/transformers/models/cosmos/modular_cosmos.py:765]: namings are weird so at least comment int he `idwt` also waht does `idwt` means?

File: src/transformers/models/cosmos/modular_cosmos.py [hunk 7]
```diff
@@ -0,0 +1,2297 @@
...
+
+class CosmosTextRMSNorm(Emu3RMSNorm):
+    pass
+
+
+class CosmosTextMLP(Emu3MLP):
+    pass
+
+
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
```
Comment (ArthurZucker) [inline src/transformers/models/cosmos/modular_cosmos.py:1344]: I think we have this already no?

File: src/transformers/models/cosmos/modular_cosmos.py [hunk 8]
```diff
@@ -0,0 +1,2297 @@
...
+    q_embed = (q_rot * cos) + (rotate_half(q_rot) * sin)
+    k_embed = (k_rot * cos) + (rotate_half(k_rot) * sin)
+
+    # Concatenate back to full shape
+    q_embed = torch.cat([q_embed, q_pass], dim=-1)
+    k_embed = torch.cat([k_embed, k_pass], dim=-1)
+    return q_embed, k_embed
+
+
+class CosmosTextAttention(LlamaAttention):
```
Comment (ArthurZucker) [inline src/transformers/models/cosmos/modular_cosmos.py:1379]: I would split cross and self like in `DiaModel`

File: src/transformers/models/cosmos/modular_cosmos.py [hunk 9]
```diff
@@ -0,0 +1,2297 @@
...
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if (input_ids is None) ^ (inputs_embeds is not None):
+            raise ValueError("You must specify exactly one of input_ids or inputs_embeds")
+
+        if self.gradient_checkpointing and self.training and use_cache:
+            logger.warning_once(
+                "`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`."
+            )
+            use_cache = False
```
Comment (ArthurZucker) [inline src/transformers/models/cosmos/modular_cosmos.py:1622]: not needed anymore!

File: src/transformers/models/cosmos/modular_cosmos.py [hunk 10]
```diff
@@ -0,0 +1,2297 @@
...
+        w_grids = math.ceil(seq_length / num_width_grid)
+        num_frames = math.ceil(seq_length / one_frame_len)
+        position_width = torch.tensor(list(range(64)) * w_grids)[:seq_length].unsqueeze(0)
+        position_height = torch.arange(w_grids).repeat_interleave(num_width_grid)[:seq_length].unsqueeze(0)
+        position_height = position_height % num_height_grid
+        position_temporal = torch.arange(num_frames).repeat_interleave(one_frame_len)[:seq_length].unsqueeze(0)
+        position_ids = torch.stack([position_temporal, position_height, position_width], dim=1).to(device)
+        return position_ids
+
+    @add_start_docstrings_to_model_forward(COSMOS_TEXT_INPUTS_DOCSTRING)
```
Comment (ArthurZucker) [inline src/transformers/models/cosmos/modular_cosmos.py:1590]: use check_model_inputs

File: src/transformers/models/cosmos/modular_cosmos.py [hunk 11]
```diff
@@ -0,0 +1,2297 @@
...
+        encoder_hidden_states: Optional[torch.Tensor] = None,
+        encoder_attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        position_ids_rope: Optional[torch.LongTensor] = None,
+        past_key_values: Optional[Cache] = None,
+        inputs_embeds: Optional[torch.FloatTensor] = None,
+        use_cache: Optional[bool] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
```
Comment (ArthurZucker) [inline src/transformers/models/cosmos/modular_cosmos.py:1604]: to remove

File: src/transformers/models/cosmos/modular_cosmos.py [hunk 12]
```diff
@@ -0,0 +1,2297 @@
...
+        cache_position: Optional[torch.LongTensor] = None,
+        **flash_attn_kwargs: Unpack[FlashAttentionKwargs],
```
Comment (ArthurZucker) [inline src/transformers/models/cosmos/modular_cosmos.py:1606]: use `TransformersKwargs`

File: src/transformers/models/cosmos/modular_cosmos.py [hunk 13]
```diff
@@ -0,0 +1,2297 @@
...
+
+        for decoder_layer in self.layers[: self.config.num_hidden_layers]:
+            if output_hidden_states:
+                all_hidden_states += (hidden_states,)
+
+            if self.config.apply_abs_pos_emb:
+                hidden_states = self.absolute_position_emb(hidden_states, position_ids)
+
+            if self.gradient_checkpointing and self.training:
+                layer_outputs = self._gradient_checkpointing_func(
```
Comment (ArthurZucker) [inline src/transformers/models/cosmos/modular_cosmos.py:1681]: use gradient checkpointing layer!

File: src/transformers/models/cosmos/modular_cosmos.py [hunk 14]
```diff
@@ -0,0 +1,2297 @@
...
+            decoder_hidden_states=outputs.hidden_states,
+            decoder_attentions=outputs.attentions,
+            encoder_last_hidden_state=encoder_hidden_states,
+            encoder_hidden_states=encoder_outputs.hidden_states if encoder_outputs is not None else None,
+            encoder_attentions=encoder_outputs.attentions if encoder_outputs is not None else None,
+            cross_attentions=outputs.cross_attentions,
+        )
+        return output if return_dict else output.to_tuple()
+
+    def _update_causal_mask(
```
Comment (ArthurZucker) [inline src/transformers/models/cosmos/modular_cosmos.py:1950]: I think not needed anymore

**Review by ArthurZucker — COMMENTED** [review]
Late review but stil! Code should be simpler with our updates!˜

**Comment by 1himan** [comment]
Is this PR still going on? I am willing to help, if updates on this are still needed. @ArthurZucker @zucchini-nlp

**Comment by zucchini-nlp** [comment]
@1himan the model got a bit low in priority and there has been almost no community request to add it. So I put it aside. Do you want to take over the PR?

**Comment by 1himan** [comment]
> @1himan the model got a bit low in priority and there has been almost no community request to add it. So I put it aside. Do you want to take over the PR?

If there were no community requests, then, no :)

**Comment by qgallouedec** [comment]
Closing this since Cosmos 1 support was deprioritized due to low community demand, and NVIDIA's Cosmos lineup is now covered by the Cosmos 3 integration merged in #46146 (cosmos3_omni / cosmos3_edge).
